Add support for user-defined operations

Rufflewind opened the pull request:
It’s not pretty …  Because of how MPI_User_function is defined, it doesn’t give us a clean way attach/access arbitrary user data.  As a result, we have to pull in libffi to hack our way through.

- Add `BorrowedDatatype`, a thin wrapper over `MPI_Datatype` (feel free to bikeshed on the name)
- Redefine `SystemDatatype = BorrowedDatatype<'static>`
- Add `DynBuffer[Mut]`
- Add `is_commutative`
- Add `UserOperation`

bsteinb commented:
Thanks. I will try to find some time to look at this tonight.

bsteinb commented on src/datatype.rs:
```diff
@@ -433,6 +450,201 @@ unsafe impl<T> BufferMut for T where T: Equivalence
+    }
+
+    /// Reborrow the buffer with a shorter lifetime.
+    pub fn as_ref<'b>(self) -> DynBuffer<'b> where 'a: 'b {
```
Better not step on `AsRef`'s toes. Why not call this `reborrow`?

bsteinb commented on src/datatype.rs:
```diff
@@ -433,6 +450,201 @@ unsafe impl<T> BufferMut for T where T: Equivalence
+    }
+}
+
+/// An mutable dynamically-typed buffer.
```
**A** mutable...

bsteinb commented on src/datatype.rs:
```diff
@@ -433,6 +450,201 @@ unsafe impl<T> BufferMut for T where T: Equivalence
+/// An immutable dynamically-typed buffer.
+///
+/// The buffer has a definite length and MPI datatype, but it is not yet known which Rust type it
+/// corresponds to.  This is like the MPI analogue of `&Any`.
```
"like" and "analogue" in the same sentence feels like too much.

bsteinb commented on src/datatype.rs:
```diff
@@ -433,6 +450,201 @@ unsafe impl<T> BufferMut for T where T: Equivalence
+/// An mutable dynamically-typed buffer.
+///
+/// The buffer has a definite length and MPI datatype, but it is not yet known which Rust type it
+/// corresponds to.  This is like the MPI analogue of `&mut Any`.
```
See above.

bsteinb commented on src/datatype.rs:
```diff
@@ -433,6 +450,201 @@ unsafe impl<T> BufferMut for T where T: Equivalence
+
+impl<'a> DynBufferMut<'a> {
+    /// Creates a mutable buffer from a mutable slice with whose type has an MPI equivalent.
+    pub fn new_mut<T: Equivalence>(buf: &'a mut [T]) -> Self {
```
Why not just `new`?

Rufflewind replied:
Dunno, I was debating this myself.  If this were rewritten using custom dynamically sized types (https://github.com/rust-lang/rfcs/issues/813) in 2030, it would be like `&BufferObj` and `&mut BufferObj`, in which case the `_mut` suffix would be necessary to distinguish the different variants.  But I guess in the meantime writing `DynBufferMut::from_mut` feels a bit redundant.

bsteinb commented on src/datatype.rs:
```diff
@@ -433,6 +450,201 @@ unsafe impl<T> BufferMut for T where T: Equivalence
+    }
+
+    /// Creates a buffer from its raw components.  The buffer must remain valid for `'a`.
+    pub unsafe fn from_raw_mut<T>(ptr: *mut T,
```
Why not just `from_raw`?

bsteinb commented on src/datatype.rs:
```diff
@@ -433,6 +450,201 @@ unsafe impl<T> BufferMut for T where T: Equivalence
+    }
+
+    /// Reborrows the buffer with a shorter lifetime.
+    pub fn as_ref<'b>(&'a self) -> DynBuffer<'b> where 'a: 'b {
```
See above.

bsteinb commented on src/datatype.rs:
```diff
@@ -433,6 +450,201 @@ unsafe impl<T> BufferMut for T where T: Equivalence
+    }
+
+    /// Reborrows the buffer mutably with a shorter lifetime.
+    pub fn as_mut<'b>(&'a mut self) -> DynBufferMut<'b> where 'a: 'b {
```
See above.

bsteinb commented on src/collective.rs:
```diff
@@ -1369,6 +1384,111 @@ unsafe impl AsRaw for SystemOperation {
+        });
+        let mut op;
+        anchor.wrapper = Some(unsafe {
+            let wrapper = ClosureMut4::new(&mut anchor.function);
```
There is something at the back of my mind about not `panic!`ing through the FFI. Does `libffi` take care of that or should this be handled here?

Rufflewind replied:
Since this is the "safe" interface of libffi, it will abort the program if the callback panics.

bsteinb commented on src/collective.rs:
```diff
@@ -1369,6 +1384,111 @@ unsafe impl AsRaw for SystemOperation {
+            let wrapper = ClosureMut4::new(&mut anchor.function);
+            op = mem::uninitialized();
+            ffi::MPI_Op_create(Some(*wrapper.code_ptr()),
+                               commute as _,
```
Wow, I did not know this was possible. Not sure I like it, weird mix of explicit cast and inferred type. No need to change this though.

Rufflewind commented on src/collective.rs:
```diff
@@ -1369,6 +1384,111 @@ unsafe impl AsRaw for SystemOperation {
+/// is still alive due to outstanding references.
+pub struct UserOperation<'a> {
+    op: MPI_Op,
+    anchor: Box<Erased + 'a>,           // keeps the internal data alive
```
Well, the closure must be boxed (it has to be immovable so that references to it remain valid). Therefore I figured I may as well simplify the type of UserOperation by erasing F, which would've otherwise been​ an unnameable closure type.

Rufflewind commented on src/datatype.rs:
```diff
@@ -60,22 +61,38 @@ pub mod traits {
-pub struct SystemDatatype(MPI_Datatype);
+/// This is a almost equivalent to a raw datatype handle, but unlike a raw handle, it is guaranteed
+/// to be a valid handle (subject to the lifetime `'a`).
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
```
Ok, `Eq` and `PartialEq` have been removed.  (`Hash` was also removed because `Hash` by itself isn't very useful.)

Rufflewind commented on src/datatype.rs:
```diff
@@ -60,22 +61,38 @@ pub mod traits {
+/// This is a almost equivalent to a raw datatype handle, but unlike a raw handle, it is guaranteed
+/// to be a valid handle (subject to the lifetime `'a`).
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct BorrowedDatatype<'a> {
```
I originally picked `DatatypeRef`, then swapped to `BorrowedDatatype` just before submitting the PR, but now I went back to `DatatypeRef` because `BorrowedDatatype` is too long to type :p  I would say `DatatypeRef` is a bit more consistent with e.g. [`std::cell::Ref`](https://doc.rust-lang.org/std/cell/struct.Ref.html).

Let me know if you want to go back to `BorrowedDatatype`.

Rufflewind commented:
- Added examples for `UserOperation`.
- Added `UserOperation::associative` and `UserOperation::commutative` as more readable aliases for `UserOperation::new(false/true, …)`
- The order of arguments for `UserOperation::new` has been changed to put the closure last.
- `downcast_mut` was renamed `downcast` and the old `downcast` has been removed (it's equivalent to `.downgrade().downcast()`)
- `downcast` takes `self` by value now and returns a slice with the same lifetime rather than a shorter lifetime

bsteinb commented:
Okay, here we go.

Rufflewind commented:
Cool, thanks!  Would you mind updating gh-pages?

bsteinb commented:
Heh, I don't even remember how to do it manually. I have Travis set-up to do it for me every time I tag a release. I guess it is time for a new release and I could even start pushing to crates.io again, now that there are no longer any dependencies on raw git repositories. I will do it by the end of the weekend.

bsteinb commented:
There we go. That was more of a struggle than I expected. Since today is a public holiday where I live, I will still consider this "by the end of the weekend".

Rufflewind commented:
Thanks! :P
